fix: Provide typed exec(query) to avoid casting results in lamdba block

bog-walk opened the pull request:
#### Description

**Summary of the change**: Introduce `execQuery()` to transaction classes for executable statements that are expected to return results, so that users will not need to cast to +/- unwrap `JdbcResult`/`R2dbcResult` in order to transform the results.

**Detailed description**:
- **Why**:

It is most likely less common than iterating over the query/statement directly, but it has always been possible to pass a `Statement` instance to `exec()` and perform a transformation on the retrieved result.

Pre-V1, with JDBC, this means a `ResultSet` was passed as the lambda block argument, so users could act directly on it to return a transformed result.

Post-V1, `exec()` still takes a statement (through the respective executable object), but the argument passed to the lambda is the core common `ResultApi`. This means users would need to cast and (for JDBC) unwrap a `JdbcResult` to fix the broken code.

- **How**:
    - Keeping the name `exec()` returns overload resolution ambiguity errors even with more restrictive parameter types
    - Instead, introduce `execQuery()` to `JdbcTransaction` and `R2dbcTransaction` 
    - Parameter `query` is not of `Query` type to ensure it allows access to all statements that return results: query, explain, returning.

---

#### Type of Change

Please mark the relevant options with an "X":
- [X] Bug fix

Affected databases:
- [X] All

#### Checklist

- [X] Unit tests are in place
- [X] The build is green (including the Detekt check)
- [X] All public methods affected by my PR has up to date API docs
- [X] Documentation for my change is up to date

bog-walk commented on documentation-website/Writerside/topics/Migration-Guide-1-0-0.md:
```diff
@@ -319,7 +328,74 @@ transaction {
+</compare>
+
+<note>
+Starting from version 1.0.0-beta-6, the original pre-version 1.0.0 code in your lambda can once again be used by replacing <code>exec(query)</code>
```
Not sure about the next version, so would put a ticket to replace this if necessary.

vnikolova replied:
I checked this with the technical writing team and we agreed it is not necessary to mention which version the changes took place and keep this document updated to match the latest release until 1.0.0.
If users need to know which version changes took place in prior to the stable release, they can refer to the _Breaking changes_ topic.

bog-walk replied:
Sure I can remove it, thanks for checking convention. In both instances mentioned in this PR, both the changes aren't mentioned in breaking changes because they're new features meant to make the migration easier. So either option can work. Maybe ideally, the original solution should be entirely replaced with the new sections, but I'll adjust it and check back with you.

bog-walk commented:
@vnikolova This is the PR I mentioned with the migration guide changes.
Please let me know if there's anything I should change about how I call out the difference between migration from 0.61.0 to 1.0.0-beta-1 vs (potential) 0.61.0 to more recent betas.

bog-walk commented:
@vnikolova Ok I've replaced the new notes sections so they don't reference the exact beta version with the change.
Following that thought, I've also removed `-beta-1` from other places in the migration guide.
Tables and comparison blocks now just show 0.61.0 versus 1.0.0. Please let me know if it's all good like this.
